### PR TITLE
use higher precision dtype for counts in batch_norm_gather_stats_with_counts

### LINKED ISSUE #70901: Accuracy problem of torch.batch_norm_gather_stats_with_counts when running_mean is half tensor
### 🐛 Describe the bug

The value of output tensor "mea" should not be zero.

```
import torch
i = torch.rand([5,4,10,10], dtype=torch.float).cuda()
c = torch.rand((150,), dtype=torch.half).fill_(500).cuda()
m = torch.rand([150, 4], dtype=torch.float).cuda()
rm = torch.rand([4], dtype=torch.half).cuda()
v = torch.rand([150, 4], dtype=torch.float).cuda()
v += 0.0001
rv = torch.rand([4], dtype=torch.half).cuda()
rv += 0.0001
mea, inv = torch.batch_norm_gather_stats_with_counts(i, m, v, rm, rv, 1.2, 0.001, c)
print(mea)
```
output:

```
tensor([0., 0., 0., 0.], device='cuda:0')
```

### Versions

PyTorch version: 1.10.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: CentOS Linux release 7.9.2009 (Core) (x86_64)
GCC version: (GCC) 5.4.0
Clang version: Could not collect
CMake version: version 3.15.5
Libc version: glibc-2.3.4

Python version: 3.6.8 (default, Nov 16 2020, 16:55:22)  [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)] (64-bit runtime)
Python platform: Linux-3.10.0-1160.15.2.el7.x86_64-x86_64-with-centos-7.9.2009-Core
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB

Nvidia driver version: 440.64.00
cuDNN version: Probably one of the following:
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] torch==1.10.1
[conda] Could not collect

### PR DESCRIPTION
**Pull request by cdzhan**
use higher precision dtype for counts and keep the same dtype with mean as used in SyncBatchNorm forward

Fixes #70901

### CODE REVIEW

**Comment by jjsjann123** [comment]
Have we looked at perf? This kernel with welford is pretty challenging on register usage.
